standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/ReplChangeManager.java

```diff
@@ -356,6 +356,9 @@ public static FileInfo getFileInfo(Path src, String checksumString, String srcCM
   // Currently using fileuri#checksum#cmrooturi#subdirs as the format
   public static String encodeFileUri(String fileUriStr, String fileChecksum, String encodedSubDir)
           throws IOException {
+    if (instance == null) {
+      throw new IOException("Uninitialized ReplChangeManager instance.");
+    }
     String encodedUri = fileUriStr;
```

Comment:
isCMFileUri logic can be changed to check if have fragments (means, # is present in url), then it should be treated as CM file URI even if checksum value is empty. This is useful for cloud to cloud replication (in future) where checksum calculation is not supported and hence CM will be disabled.

Reply:
When we create ReplCopyTask for FunctionHandler during load, we know that this is going to a set of direct files. I think a better fix there would be to tell ReplCopyTask that this is a direct set of files (instead of _files) through a flag in constructor instead of relying upon annotations in URL. We may want to change the structure of URL (jar url) later or interpret it differently. This will help in cloud to cloud replication, which may have require different structure of URL.

Given that there are other problems that we have to fix for cloud to cloud other than this and also the uncertainty about what exactly will work with cloud, we may adopt a different strategy here. So for the time being I will leave this untouched.

Reply:
I agree that we need not fix it right away as we currently enforce to enable CM for replication to work.
But, in HDP 2.6, we already support replication even when CM is disabled which is broken in HDP 3.1. This is due to difference in implementation of isCMFileUri method in both these versions.